### PR TITLE
Altitude Control: Console, Skills, TGUI, Shuttle, OB, And Hijack changes.

### PR DESCRIPTION
**Pull request by Moonshanks**
# About the pull request
This pull request adds a feature via the /machinery/computer/attitude_control_console.

The USS Almayer can now be "moved" between three different levels of orbit/proximity to the AO:

High,
Optimal,
Low

Each level comes with changes to the duration of transport timers, time to transport during hijack, and OB cooldown timer, equal to the GLOB.ship_alt value.

High makes OB cooldowns take 200 seconds longer, dropship transport times take 50% longer, and the dropship hijack transport time take 50% longer

Optimal, the default which will stay in place if no one touches the console leaves everything with their default times and cooldowns,

Low, OB cooldowns take 200 shorter, dropship transport times are 50% shorter, the dropship hijack transport time is 50% shorter.

While in Low orbit the Almayer will periodically shake (50% chance to shake every 30 seconds), the Almayer's thrusters will also start building up heat as they battle with making minor adjustments due to the dangerous proximity to the AO. They will gain 10% of their max heat each 30 seconds, once they reach 100% the ship will be forced into the highest orbit to cool off, and cool off slower than normal (5% every 30 seconds) until its orbit is changed.

While in Optimal or High orbit the Almayer thrusters will cool off by 10% per 30 seconds.

The Almayer's orbit may only be changed by those with a "navigations" skill of 1. (only the CO and Synth -- EDIT: XO now has the skill too -- currently but I may add a dedicated RP role for this mechanic later down the line). The orbital level may only be changed every 90 seconds and when it changes the ship will shake violently causing every mob on the z level to fall over.

This PR does not place the Altitude Control Console on the map, so currently, these features don't do anything within a normal round unless staff spawns in the console, however I will be uploading a mapping PR changing the Astronavigations deck if this PR is accepted.

Planned for the future but not yet approved connected to this PR is the "Navigations Officer" the highest auxiliary support personnel with skills the same as an SO, sans a 1 in Navigations, and a 1 or 2 in piloting. The idea for this future role would be set out in the PR, but it would represent a mainly fluff officer role that was unable to deploy under normal SoP.

# Explain why it's good for the game

This PR is a non-intrusive way to add more nuanced gameplay mechanics to COs, especially on the quieter rounds when they aren't swamped by OW duties.

The way the PR is currently designed it doesn't effect any gameplay balance if it isn't used. If a CO however chooses to use it they have to pay some level of attention to it or they will overheat the engines and cause their transport times to be lengthened.

It's a relatively simple way to add more complexity to CIC, and give the CO/Synth more stuff to do to gain a slight edge.

I have been able to test everything other than the hijack time increase. However, the line of code handling the hijack time increase is one line long. Everything else is confirmed as working and the common bugs this could create have been tested for and not found (transport shows the right time when the time is modified, OB shows the right cooldown time, the cooling can't drop the heat% below 0 nor above 100, the TGUI works without issues, the console can only be used by those with the right skills, and the knockdown effects all mobs on the Almayer not just humans).

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

https://www.youtube.com/watch?v=-cbnqNtKyCY - video showing the CO and Synth using the console, with the knockdown effect and arrest radio announcements.

https://youtu.be/Qd37iM-4FrQ - video of the overheat function and the ship shaking due to low orbit

https://www.youtube.com/watch?v=EWLCDZp-9iI - video of the ship being left on low orbit for too long, and what happens when the engines overheat

https://www.youtube.com/watch?v=u_ErqfU-nus - video showing the orbital distance effecting the transport time

https://www.youtube.com/watch?v=j687yqlWLT8 - video showing high orbit effecting the OB cooldown time.

</details>

# Changelog

:cl:
add: added altitude control console and related mechanic
add: added the 'navigations' skill for using the console and applied it to the CO/Synth
balance: added a mechanic for COs to reduce transport and OB cooldown times, and increase hijack transport times
/:cl:

### CODE REVIEW

**Comment by Moonshanks** [comment]
Oh uh before I forget. THis PR is 90% @harryob and @BeagleGaming1 work with handholding me through every single bit in such a painful manner.

I cannot thank them enough for their sacrifice.

Some advice was also provided by @paulrpg. 

Thanks nerds.

**Comment by kiVts** [comment]
changelog name/credits? like, after first 🆑

**Comment by Moonshanks** [comment]
On request from discord, XO got the skill too. While it's _sort of__ outside of the scope for this PR we can always change the skills and who has them if a extra role is added at a later date.

**Comment by realforest2001** [comment]
Can you add a log for altitude changes please. log_game or message_admins should be alright

**Comment by Moonshanks** [comment]
> Can you add a log for altitude changes please. log_game or message_admins should be alright

on it.

Spotted a few mistakes I made with the CAS timer so fixing that too.

File: code/modules/cm_marines/altitude_control_console.dm [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+			if(SHIP_ALT_LOW)
+				temperature_change = HEATING
+			if(SHIP_ALT_MED)
+				temperature_change = COOLING
+			if(SHIP_ALT_HIGH)
+				temperature_change = COOLING
+	GLOB.ship_temp += temperature_change
+	if(GLOB.ship_temp < 0)
+		GLOB.ship_temp = 0
+		return
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:64]:
```suggestion
	GLOB.ship_temp = Clamp(GLOB.ship_temp += temperature_change, 0, 120)
```

File: code/modules/cm_marines/altitude_control_console.dm [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+	if(prob(50))
+		return
+	if(GLOB.ship_alt == SHIP_ALT_LOW)
+		for(var/mob/living/carbon/current_mob in GLOB.living_mob_list)
+			if(!is_mainship_level(current_mob.z))
+				continue
+			shake_camera(current_mob, 10, 1)
+		ai_silent_announcement("Performing Attitude Control", ";", TRUE)
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:72]: altitude

File: code/modules/cm_marines/altitude_control_console.dm [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+		if("high_alt")
+			change_altitude(user, SHIP_ALT_HIGH)
+			. = TRUE
+	message_admins("[key_name(user)] has changed the ship's altitude to [action].")
+
+	add_fingerprint(usr)
+
+/obj/structure/machinery/computer/altitude_control_console/proc/change_altitude(mob/user, new_altitude)
+	if(TIMER_COOLDOWN_CHECK(src, COOLDOWN_ALTITUDE_CHANGE))
+		to_chat(user, SPAN_WARNING("the engines are not ready to burn yet."))
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:118]:
```suggestion
		to_chat(user, SPAN_WARNING("The engines are not ready to burn yet."))
```

File: code/modules/cm_marines/altitude_control_console.dm [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	if(GLOB.ship_alt == new_altitude)
+		return
+	GLOB.ship_alt = new_altitude
+	TIMER_COOLDOWN_START(src, COOLDOWN_ALTITUDE_CHANGE, 90 SECONDS)
+	for(var/mob/living/carbon/current_mob in GLOB.living_mob_list)
+		if(!is_mainship_level(current_mob.z))
+			continue
+		current_mob.apply_effect(3, WEAKEN)
+		shake_camera(current_mob, 10, 2)
+	ai_silent_announcement("Attention: performing high-G maneuver", ";", TRUE)
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:129]:
```suggestion
	ai_silent_announcement("Attention: Performing high-G manoeuvre", ";", TRUE)
```

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 5]
```diff
@@ -84,6 +84,11 @@
 		if(istype(equipment, /obj/structure/dropship_equipment/fuel/cooling_system))
 			recharge_duration = recharge_duration * SHUTTLE_COOLING_FACTOR_RECHARGE
 
+	//factors in the distance to the AO when in transit
+if(!is_set_flyby)
```
Comment (harryob) [inline code/modules/shuttle/computers/dropship_computer.dm:88]: indentation

**Review by harryob — CHANGES_REQUESTED** [review]

**Comment by harryob** [comment]
> changelog name/credits? like, after first 🆑

@kiVts unnecessary, you only need that to override it being your github username

File: code/modules/cm_marines/altitude_control_console.dm (hunk 2 above)
Comment (harryob) [reply code/modules/cm_marines/altitude_control_console.dm:72]: oh yeah he’s right

File: code/modules/cm_marines/altitude_control_console.dm [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+//Has the ships temperature set to 0 on startup, sets the global default var to med
+
+GLOBAL_VAR_INIT(ship_temp, 0)
+GLOBAL_VAR_INIT(ship_alt, SHIP_ALT_MED)
+
+/obj/structure/machinery/computer/altitude_control_console
+	icon_state = "overwatch"
+	name = "altitude control console"
+	desc = "placeholder desc"
```
[the remark this answers is not in the document]
Comment (Moonshanks) [reply code/modules/cm_marines/altitude_control_console.dm:22]: Add the desc on it.

File: code/modules/cm_marines/altitude_control_console.dm (hunk 1 above)
Comment (Moonshanks) [reply code/modules/cm_marines/altitude_control_console.dm:61]: I thought there was a clamp function somewhere.

File: code/modules/shuttle/computers/dropship_computer.dm (hunk 5 above)
Comment (Moonshanks) [reply code/modules/shuttle/computers/dropship_computer.dm:88]: fixed. Testing now.

Comment (Moonshanks) [reply code/modules/shuttle/computers/dropship_computer.dm]: Works. Resolving this.

**Comment by Kitsunemitsu** [comment]
I play a lot on lowpop and for contingency I would also give it to the CE. 
CE makes sense to have it and this makes 2 (later 3) non-WL roles to have it, leaving these features open for use at 6-11am GMT.

An analogue would be the OB. You don't need the OB to win an op, but it is a useful tool, which is why there are plenty of contingencies on who can load the OB (SO, XO, CO, Synth, CE, OT, MT) so that you're not locked out of these features on lowpop.

File: code/modules/cm_marines/altitude_control_console.dm [hunk 7]
```diff
@@ -0,0 +1,132 @@
+//Defines what to times the OB and transport shuttle timers by based on the altitude
+#define SHIP_ALT_LOW 0.5
+#define SHIP_ALT_MED 1
+#define SHIP_ALT_HIGH 1.5
+
+//List of available heights
+GLOBAL_VAR_INIT(ship_alt_list, list("Low Altitude" = SHIP_ALT_LOW, "Optimal Altitude" = SHIP_ALT_MED, "High Altitude" = SHIP_ALT_HIGH))
+
+//Handles whether or not hijack has disabled the system
+GLOBAL_VAR_INIT(alt_ctrl_disabled, 0)
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:10]:
```suggestion
GLOBAL_VAR_INIT(alt_ctrl_disabled, FALSE)
```

**Comment by Moonshanks** [comment]
> I play a lot on lowpop and for contingency I would also give it to the CE. CE makes sense to have it and this makes 2 (later 3) non-WL roles to have it, leaving these features open for use at 6-11am GMT.

I hesitated to give it to the XO because of the console's ability to make everyone fall over every 90 seconds if spammed.
I'd rather not give it to anyone other than the top of the command structure and a specialised role for it.

I don't see a good reason to give it to the CE other than to increase it's usage, and the way this PR is designed this console is supposed to be something that isn't used all the time as optimally as possible. It's supposed to be an added extra for command to do that if they don't touch won't do anything, rather than a new mechanic that the CE gets assigned to doing all round because it provides technically optimal gameplay.

**Comment by Moonshanks** [comment]
That being said there is nothing stopping you from giving the skill to CE in a PR once it is up. I just think doing so goes against the design idea of this PR.

File: code/modules/cm_marines/altitude_control_console.dm [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+	icon_state = "overwatch"
+	name = "Altitude Control Console"
+	desc = "The A.C.C console monitors, regulates, and updates the ships attitude and altitude in relation to the AO. It's not rocket science."
+
+/obj/structure/machinery/computer/altitude_control_console/attack_hand()
+	. = ..()
+	if(!skillcheck(usr, SKILL_NAVIGATIONS, SKILL_NAVIGATIONS_TRAINED))
+		to_chat(usr, SPAN_WARNING("A window of complex orbital math opens up. You have no idea what you are doing and quickly close it."))
+		return
+	if(GLOB.alt_ctrl_disabled == 1)
```
Comment (harryob) [inline code/modules/cm_marines/altitude_control_console.dm:32]:
```suggestion
	if(GLOB.alt_ctrl_disabled)
```

File: code/modules/shuttle/computers/dropship_computer.dm [hunk 9]
```diff
@@ -249,6 +254,7 @@
 	dropship.is_hijacked = TRUE
 
 	hijack.fire()
+	GLOB.alt_ctrl_disabled = 1
```
Comment (harryob) [inline code/modules/shuttle/computers/dropship_computer.dm:257]:
```suggestion
	GLOB.alt_ctrl_disabled = TRUE
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Review by morrowwolf — APPROVED** [review]
Architect approval.

Given the console is not actually in the gameplay loop yet with this PR and instead this just sets the groundworks I am fine with getting it through and messing with numbers in the future.

As a note though, when the console gets added in whatever form the numbers will have to be looked at and downsides to being in low orbit will need to be added.

**Comment by cm13-github** [comment]
#### Test Merge Deployed

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: f96439dea585696374f0f74eeaac0189748efd83
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: 31dd7bac120691339519407fbaebb60be862f3c1

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: f96439dea585696374f0f74eeaac0189748efd83
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: c331f6e0c309e380bfe1c0b92001ed78b994ddcf

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: c121ee6752a4af8076dbce9aa63651c5802b098a
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: b951f3dbe2c5b84502bedc2316789f1b3ebf9abd

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: cb2716a6f3a07dfcd234d64d21a54d46e6955408
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: 30b545f159c2cd382c9a9d84d80535401d456608

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: cb2716a6f3a07dfcd234d64d21a54d46e6955408
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: 2a28b4aba52c786f6b9a56978af6f6e1ec5d6867

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: 42bde1e0ccfeca19fcfcdd996a07e7ad42e1d623
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: db03b5326e26106c8c582b792adbd765dad1dd0f

##### Comment

</details>

**Comment by cm13-github** [comment]
#### Test Merge Updated

<details><summary>Details</summary>

##### Server Instance
CM13 Live Instance

##### Revision
Origin: 46e938a9a59b7cf7b7f80392f1de3a6c3ed1f015
Pull Request: ff5a20540cf1a78ea5cd0ea57fcd4c424b05af25
Server: 9ff946aa5d61af4999e86d21ab1ab48e99fa9c34

##### Comment

</details>

**Review by harryob — APPROVED** [review]

**Comment by cm13-github** [comment]
#### Test Merge Removed

**Comment by CocaColaTastesGood** [comment]
@Moonshanks 
"message_admins("[key_name(user)] has changed the ship's altitude to [action].")"
might be a good idea to sanitize action or make a default case that returns if not one of the 3 options.

**Pull request merged**